### DISCUSSION TITLE
Request: Download from netu-Player

### DISCUSSION THREAD

**Discussion by AitschPi** [aclap-dev/video-downloadhelper#824] [Help and Q&A]
### Your question

Some websites integrate videos using the player of netu. These videos where not detected by the addone. 

Example: https://maxi.show/it/comedy/175-kendi-dusen-aglamaz.html

### Addon & CoApp version and details

version: 9.0.2.3
target: mozilla
channel: stable
lang: de
coapp: {"found":true,"path":"/Applications/net.downloadhelper.coapp.app/Contents/MacOS/vdhcoapp","version":"2.0.19","new_version":false}
license: {"accepted":true,"email":"<removed>","key":"adfcf72f845f4478"}
platform: aarch64 mac
UA: Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:126.0) Gecko/20100101 Firefox/126.0
{
    "valid_license_message_has_been_displayed": true,
    "used_sidebar_button": true,
    "used_popup_button": true,
    "download_directory": "/Users/aitschpi/dwhelper",
    "never_show_no_incognito_msg_again": true,
    "view_options": {
        "all_tabs": false,
        "low_quality": false,
        "sort_by_status": true,
        "sort_reverse": false,
        "show_button_clean": true,
        "show_button_clean_all": false,
        "show_button_convert_local": false,
        "hide_downloaded": false
    },
    "use_sidebar": false,
    "open_count_store": 64,
    "successfull_dl": 45
}

**Comment by Mattmjs** [comment]
I'm not the dev of the software but a user like you. I looked at the video on that link and no detection like you say.
I tried opening the developer tools to see what's going on. But they won't allow the developer tools to be open while the
video is playing. This is a tactic some sites deploy to stop people from fishing around in there. 

Ok no problem, there is a way to record network activity using the developer tools without having it open. And the website doesn't
know any better. I see why there is no detection, there is no streaming manifest to be found. It seems there is some stealthy method
of streaming the video uses. I don't know if downloading will be possible for this site. I did find an m3u8 link after saving the response 
from the embed video link in a text file. I used this in ffmpeg but there was an error:

[tcp @ 000001b34be46dc0] Connection to tcp://4fw4gd.cfglobalcdn.com:443 failed: Error number -138 occurred
https://4fw4gd.cfglobalcdn.com/secip/1/861rQM940fF8R1fZDCdglg/OTQuMjUuMTcwLjI2/1606597200/hls-vod-s03/flv/api/files/videos/2018/08/01/153311550983uua.mp4.m3u8: Error number -138 occurred

Perhaps @paulrouget  would have more insight on what this site is doing and whether it'll work or not.

**Reply by AitschPi** [reply]
I also tried the developer tools… But it seems like Grim‘s „The Hare and the Hedgehog“. This player is the only one I know that won‘t work with the downloader (yet).

**Reply by GrampaWildWilly** [reply]
I do not work for VDH.  I'm just another user, like @Mattmjs Matt.

I captured a few seconds of playback using about:logging.  What I got was a Firefox activity log of almost half a G in barely 10 seconds of playback.  The activity log had many (I didn't determine how many) HLS master manifests but the handful I ran through ffprobe all said they were 55 seconds long.  They each had perhaps 16 Programs (in ffprobe parlance).  There appeared to be 2 audio Streams scattered about among the multiple video Streams.  There were some video Streams identified as H264, which would be for MP4 files.  There were some video Streams identified as VP9, which would be for MKV files.  All of the URLs for the manifests I found looked like this content was being served from YouTube.  I tried downloading (with ffmpeg) what I found in one of these manifests but it appeared to be the advertisement of 55 seconds that VDH is recognizing on this page.  It certainly wasn't what I saw in the video player on the web page for the 10 seconds I let it play.

I don't know how I would find the manifest for the movie itself.  I have to assume it was in there.  But finding it is prohibitively difficult.  You can't tell by looking at the URL what it's for.  You would have to painstakingly run ffprobe on each manifest you find.  I have a life to live.  I can't run some untold number of ffprobe invocations on manifests that might or might not be the one for the movie.

My suspicion is that this site generates master manifests in a chain.  Each link in the chain covers some short duration of the movie.  If you could find the first link in the chain, there is a possibility, no more than a possibility, that you could start ffmpeg on that manifest & as the chain was fed out to ffmpeg, it would record the whole movie.  But it seems to me that this site has gone to great lengths to make it difficult to download its content.

On the other hand, like I said, this content appears to be hosted on YouTube.  If you can find the thing on YouTube, VDH knows very well how to download things from YouTube.

I am certainly giving up on this one at this point.  You can do as you please, but if you manage to figure this one out, do share your findings here.

**Reply by GrampaWildWilly** [reply]
By the way, this is not an issue about the netu player.  netu player, jwplayer, BBC iplayer, Brightcove player, these are all distractions.  The player is not an issue.  VDH detects network activity.  It detects the same network activity the player embedded in the web page detects.  VDH is oblivious to what player is on the page.  Both VDH & the embedded player are detecting network activity.  That's at a low level, below the embedded player.  The serving web site sends data & your client web browser receives it.  Then either the embedded player on the web page, or VDH, or both detect whatever it is & either do or don't properly interpret what they've been fed.  It is possible that you will find another web site that uses the netu player & VDH will work fine & we will be able to work on the site with the Developer Tools open.  It's not about the player.  It's about the way the web site is architected.

**Reply by GrampaWildWilly** [reply]
You might be interested in these:

https://www.youtube.com/@KendiDusenAglamazTRT
https://www.youtube.com/watch?v=iNMKK3Ya4r8

I didn't bother pushing those web pages through a translator so I don't know for certain what I'm looking at.  But I do believe the second of those URLs is for the particular episode of this series (I'm guessing it's a TV series) that we have all been having trouble with on that maxi.show site.  VDH was quite happy to download the video at that second URL.  This is what I got:

![#01](https://github.com/aclap-dev/video-downloadhelper/assets/75588964/bda18ff2-6af8-4d48-a4d7-33bd9f0b70c4)

And here it is playing in VLC:

![#02](https://github.com/aclap-dev/video-downloadhelper/assets/75588964/2b513503-d9c2-4ccf-b781-771e62a200df)

You can see it looked just fine.  You'll have to take my word for it that it also sounded just fine.  But I have no idea what they were saying.  It's a language I don't understand.  Turkish?  That's one reason I just skimmed through this quickly, sampling it at intervals to make sure it was all there.  It even looked to have the complete closing credits.

My advice is to quit wrestling with this useless maxi.show web site that seems determined to thwart all our efforts.  Go to YouTube where you can find the exact same thing, where VDH works perfectly.

I suppose I should specify that I'm running Windows 11 Pro 64-bit version 23H2 build 22631.3672, Firefox 126.0.1 64-bit, licensed VDH 9.0.2.3a1 beta, CoApp 2.0.19.  If you run that environment, you will successfully download any of these episodes you might care to.

**Reply by GrampaWildWilly** [reply]
I suppose if you want the Italian subtitles you get from maxi.show, you're out of luck.  On the other hand, if you're really motivated, there are applications out there that will generate Italian subtitles from the Turkish dialog.  I can't vouch for their accuracy or their ease of use.  But I know they exist.  I just haven't used any of them.

**Comment by Mattmjs** [comment]
You can get a better look in the developer tools / Network by switching on a setting in ublock origin.
In the filters list tick ublock filters - annoyances. Then the video won't complain that the developer tools are open.

Here is an example that was posted on stackoverflow **https://netu.tv/view_page.php?pid=4**
If you save the response from the embedded player and view it in a text editor. there is whole lot of code.

![netu player](https://github.com/user-attachments/assets/f0621da2-3e82-45ac-a037-dda6dc38d90b)

Use the find function to look for m3u8, there is one url that it finds. But it has the same error that I saw for the other site up the top
in ffmpeg.

**[tcp @ 000001eb59f56d80] Connection to tcp://4fw4gd.cfglobalcdn.com:443 failed: Error number -138 occurred
https://4fw4gd.cfglobalcdn.com/secip/1/861rQM940fF8R1fZDCdglg/OTQuMjUuMTcwLjI2/1606597200/hls-vod-s03/flv/api/files/videos/2018/08/01/153311550983uua.mp4.m3u8: Error number -138 occurred**

In yt-dlp it says this
**[generic] Extracting URL: https://4fw4gd.cfglobalcdn.com/secip/1/861rQM940fF8R1fZDCdglg/OTQuMjUuMTcwLjI2/1606597200/hls-vod...11550983uua.mp4.m3u8
[generic] 153311550983uua.mp4: Downloading webpage
ERROR: [generic] Unable to download webpage: <urllib3.connection.HTTPSConnection object at 0x000001C80116E100>: Failed to establish a new connection: [WinError 10061] No connection could be made because the target machine actively refused it (caused by TransportError('<urllib3.connection.HTTPSConnection object at 0x000001C80116E100>: Failed to establish a new connection: [WinError 10061] No connection could be made because the target machine actively refused it'))**

If it's possible to workaround it, it may possibly work.

**Reply by lonemadmax** [reply]
For the privacy paranoid: those links contain the IP and time of your request, maybe other data. Consider that before posting them. They are also probably time limited.

For the video from netu.tv, one URL I get for the manifest is https://t9lr3d.cfglobalcdn.com/silverlight/secip/7755/12479/FpMUkRMQofKcKpEsNqk6uw/OTAuMTYyLjE0Ni4xNDc=/1748985067/hls-vod-s002/flv/api/files/videos/2016/03/01/14568685065eb92, surely a different video now, as it has a different upload time.

Tried getting it with curl and I get 404 or no response unless I add the right `Referer` and `Origin` headers, which I don't know if you can do with ffmpeg. The lack of response might be due to my connection, so maybe just Origin will do. On top of that, the response has `Content-Type: application/octet-stream` and `X-Content-Type-Options: nosniff` headers. So unless the addon ignores those and sniffs the content of everything, it won't know it's a manifest. Although curl gave me a nice file, the response in firefox's dev tools is base64-encoded, if it makes any difference.

The video fragments are "manifest URL".mp666/Frag-1-v1-a1, where that first 1 goes up sequence. These are again blocked without the right headers, and for some added fun, respond as `text/x-html`, `nosniff`.

If you get the right manifest URL and origin header and the stars align, this shell script *may* get you a video file:

```bash
#!/bin/bash
set -e

if [ $# -ne 3 ]; then
  >&2 printf "%s URL ORIGIN FILE\n" "$0"
  exit 1
fi

UA="User-Agent: Mozilla/5.0 (X11; Linux x86_64; rv:138.0) Gecko/20100101 Firefox/138.0"
PLAYER=https://netuplayer.top

URL="$1"
PLAYER="$2"
FILE="$(realpath $3)"

if [ -f "${FILE}" ]; then
  >&2 printf "%s already exists\n" "${FILE}"
  exit 2
fi

TMPWD="$(mktemp -d)"
pushd "${TMPWD}"
curl -L -H "${UA}" -H "Origin: ${PLAYER}" -H "Referer: ${PLAYER}" "${URL}" -o manifest.m3u8
sed '/^#/d' manifest.m3u8 | while read -r url; do curl -L -H "${UA}" -H "Origin: ${PLAYER}" -H "Referer: ${PLAYER}" -H 'Accept: video/mp4' "${URL%/*}/${url}" -o "${url##*/}"; done
sed -i 's,.*/,,' manifest.m3u8
ffmpeg -protocol_whitelist file -allowed_extensions ALL -extension_picky 0 -i manifest.m3u8 -c copy "${FILE}"
popd

rm -r "${TMPWD}"
```

For a site where I use it, I have to pipe curl's output through `gunzip`, but the responses for the one from netu.ca were not gzip-encoded, so YMMV.

**Reply by Mattmjs** [reply]
@lonemadmax , you can add referer and origin headers to ffmpeg I think. 
So if this link is used https://netu.tv/view_page.php?pid=4 what would the referer and origin headers be.

**Reply by lonemadmax** [reply]
It was https://netu.ca for me, as I was redirected. You can see all the headers if you click in the request in dev tools.

**Reply by Mattmjs** [reply]
Ok I see what it is, I can add the referer but I don't think there is a origin parameter for ffmpeg.

**Reply by Mattmjs** [reply]
> For the video from netu.tv, one URL I get for the manifest is https://t9lr3d.cfglobalcdn.com/silverlight/secip/7755/12479/FpMUkRMQofKcKpEsNqk6uw/OTAuMTYyLjE0Ni4xNDc=/1748985067/hls-vod-s002/flv/api/files/videos/2016/03/01/14568685065eb92, surely a different video now, as it has a different upload time.

That's the same link I've been looking at, the link is the playlist for all the fragments. VDH uses ffmpeg to handle the download and
it won't handle such links unless it has the .m3u8 extension. I don't know how to use cURL to get all the fragments from the playlist
but I used it on one of the fragment links. This saved a 19 second piece of the video.

**Reply by lonemadmax** [reply]
> That's the same link I've been looking at, the link is the playlist for all the fragments. VDH uses ffmpeg to handle the download and

Can't be the same, as it includes my IP and time of request + 6 hours. It's probably the same up to /12479/ and from /hls-vod, with the middle part being different.

> it won't handle such links unless it has the .m3u8 extension

Or the correct mimetype. But that is also not provided.

> I don't know how to use cURL to get all the fragments from the playlist

The sample script extracts those from the downloaded manifest (the `sed '/^#/d' manifest.m3u8` part) and feeds them to curl one by one.

**Comment by Froktime** [comment]
If anyone wants to download on netu then use File Centipede. It works very well with Netu, however there are limitations on usage. To extend usage, you have to buy premium
